incubator/virtualcluster/doc/customresource-syncer.md

```diff
@@ -0,0 +1,185 @@
+
+### Multi-tenancy Infrastructure Utility Object Construction
+
+CR controller is responsible for instantiating CR specific Mccontroller, Patroller and Listener from following packages:
+
+```
+ Import (
+ "sigs.k8s.io/multi-tenancy/incubator/virtualcluster/pkg/util/mccontroller"
+ "sigs.k8s.io/multi-tenancy/incubator/virtualcluster/pkg/syncer/patrol"
+ "sigs.k8s.io/multi-tenancy/incubator/virtualcluster/pkg/util/listener")
+```
+
+Foo Mccontroller can be constructed as:
+
+```
+multiClusterFooController, err := mccontroller.NewMCController(&v1alpha1.Foo{}, &v1alpha1.FooList{}, c,
+ mc.WithMaxConcurrentReconciles(constants.DwsControllerWorkerLow), mc.WithOptions(options.MCOptions))
+```
+
+Foo Protroller can be constructed as:
+
+```
+fooPatroller, err := patrol.NewPatroller(&alpha1.Project{}, c, pa.WithOptions(options.PatrolOptions))
+```
+
+Foo Listener is constructed as:
+
+```
+func (c *controller) GetListener() listener.ClusterChangeListener {
+        return listener.NewMCControllerListener(c.multiClusterFooController, mc.WatchOptions{})
+}
+```
+
+### CR Client Construction
+
+Current Multi-tenancy syncer uses client-go library to build shared client and informer for all standard K8s resources. However client-go client cannot embed CR client.  CR controller needs to construct CR client using Restful config of super cluster and tenant virtual cluster.
+
+Following code shows how to constructing CR Client and informer for super cluster:
+
+```
+import ("sigs.k8s.io/controller-runtime/pkg/client"
+       "sigs.k8s.io/controller-runtime/pkg/cache")
+
+ 
+ superFooClient, err := client.New(config.RestConfig, client.Options{})
+ 
+superFoocache, err = cache.New(config.RestConfig, cache.Options{})
+superFooInformer, err := c.superFoocache.GetInformer(context.Background(), &alpha1.Foo{})
+  
+```
+
+CR cache needs to be bootstrapped, We can put the bootstrap logic in DWS or UWS initialization function as:
+
+```
+func (c *controller) StartDWS(stopCh <-chan struct{}) error {
+go superFoocache.Start(stopCh)
+}
+```
+
+
+For Virtual Cluster CR Client, we can either get informer from CR specific Mccontroller, or construct client using tenant virtual cluster Restful config.
+
+```
+ clusterInformer, err := c.multiClusterFooController.GetCluster(cluster).GetInformer(&alpha1.foo{})
+ 
+ vcRestConfig := c.MultiClusterController.GetCluster(vclusterName).GetRestConfig()
+ virtualFooClient, err := client.New(vcRestConfig, client.Options{})
+```
+
+### CR Scheme Building
+
+Custom Resource Scheme installation happens in Init function of each CR Syncer package.  Controller runtime scheme builder is used here.
+
+```
+import ( 
+    "k8s.io/apimachinery/pkg/runtime"
+    "k8s.io/apimachinery/pkg/runtime/schema"
+    )
+var SchemeGroupVersion = schema.GroupVersion{
+        Group:   "foo.example.com",
+        Version: "v1alpha1",
+}
```

Comment:
Any Custom Resource developer should have already exposed these funcs and we don't need to replicate them for example we could just import this package https://github.com/kubernetes/sample-controller/blob/master/pkg/apis/samplecontroller/v1alpha1/register.go#L28 and the instead of exporting our own `SchemeBuilder` and  `localSchemeBuilder` and this file should just need:

```
import (
    "k8s.io/sample-controller/pkg/apis/samplecontroller/v1alpha1"
    "k8s.io/apimachinery/pkg/runtime"
)

var scheme = runtime.NewScheme()
func init() {
    v1alpha1.AddToScheme(scheme)
}
```

Reply:
agree